### PR TITLE
llama : more tokenizer fixes

### PR DESCRIPTION
**Pull request by ggerganov**
**TLDR: after this PR, devs should no longer add in the user code a leading space to the text they need to tokenize**

In other words, if your code uses `llama.cpp` library for tokenization and is doing this:

```c++
     s.insert(0, 1, ' '); // add a space to match OG LLaMA tokenizer
     p = ::llama_tokenize(ctx, s, add_bos); 
```

It should now just do this:

```c++     
     p = ::llama_tokenize(ctx, s, add_bos); 
```

---

note: by "user" below I mean developers. I.e. users of `llama.cpp` library

This PR should not change the tokenization results from `master` - it just improves the tokenizer API and fixes the decoding logic and the reason for the leading space in the detokenized results. It adds a few more tokenizer tests and a helper Python script to compare the behaviour of the 2 implementations. Also confirmed that `wiki.test.raw` tokenizes to `335688` tokens in exactly the same way as the OG Python tokenizer and fixed a minor bug in `perplexity` tool.

The main mistake we have been doing is to leave to the user to add the leading whitespace before calling `llama_tokenize()`. For example:

https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/main/main.cpp#L198-L201

https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/server/server.cpp#L288-L290

https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/server/server.cpp#L310-L314

This leading space is indeed needed since without it, we do not get the same results as the OG Python tokenizer. It does not matter if the provided text has one or more on none leading spaces - we **always** must prefix it with a leading space. Therefore, this has to be part of the internal implementation to avoid the user (developer) forgetting to do it:

https://github.com/ggerganov/llama.cpp/blob/e4324cbd4d367fb9ddaaef98c0929acebbf22954/llama.cpp#L3391-L3398

For example, we have technically always had a mistake in the `perplexity` tool because we haven't added the whitespace there:

https://github.com/ggerganov/llama.cpp/blob/04f4b1eb10f3e25750ca3e530265ce2841730e6b/examples/perplexity/perplexity.cpp#L190-L197

This shouldn't lead to any dramatic changes in the results, but it's better to have it the right way. With the current PR I verified that the produced tokens by `llama.cpp` match exactly the ones from Python for `wiki.text.raw`:

```bash
# tokenize with Python
python3 tests/test-tokenizer-0-llama.py ../llama2/ --fname-tok ./build/wikitext-2-raw/wiki.test.raw

# tokenize with llama.cpp
make -j && ./bin/test-tokenizer-falcon.0 ../models/ggml-vocab-llama.gguf ./wikitext-2-raw/wiki.test.raw

# diff
diff wikitext-2-raw/wiki.test.raw.tok wikitext-2-raw/wiki.test.raw.tokcpp 
```

---

I've also update the C-style API with better name of the de-tokenizer function: `llama_token_to_str -> llama_token_to_piece`:

https://github.com/ggerganov/llama.cpp/blob/e4324cbd4d367fb9ddaaef98c0929acebbf22954/llama.h#L384-L393

The `common` lib now provides a `llama_detokenize()` function that correctly detokenizes a vector of tokens into text, without adding a leading space at the start:

https://github.com/ggerganov/llama.cpp/blob/e4324cbd4d367fb9ddaaef98c0929acebbf22954/common/common.h#L119-L137

The logic is that we have to drop the leading space of the first non-BOS token when detokenizing:

https://github.com/ggerganov/llama.cpp/blob/e4324cbd4d367fb9ddaaef98c0929acebbf22954/common/common.cpp#L750-L768

This function should behave exactly the same as the Python's `tokenizer.decode()`.

Notice that in `main` we still get a leading space inserted when we print out the input prompt. This is because we decode the generated text token by token - i.e. piece by piece. So we don't have the information which token is first. We can add a state to `main.cpp` to track this, but I think it is not worth it.

Another small bug fix is that we now correctly tokenize the empty string `''` to `[1]` with `bos == true` and to `[]` with `bos == false`

Please give this branch a thorough test to make sure I haven't made some mistake and lets hope we have finally resolved all tokenizer related issues :)

---

TLDR: after this PR, devs should no longer add in the user code a leading space to the text they need to tokenize as we have been doing up until now

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -3373,22 +3371,25 @@ struct llm_tokenizer_bpe {
...
+    }
+
     if (raw_text.empty()) {
         return output;
     }
 
-    if (bos && vocab.special_bos_id != -1) {
-        output.push_back(vocab.special_bos_id);
-    }
+    raw_text = " " + raw_text;
```
Comment (klosax) [inline llama.cpp:3385]: We should have this optional like the `add_bos` parameter.

Suggestion:
`static std::vector<llama_vocab::id> llama_tokenize_internal(const llama_vocab & vocab, std::string raw_text, bool prepend_bos, bool prepend_whitespace) {`

Comment (ggerganov) [reply llama.cpp]: I moved it to the SPM branch. It is not optional - without this, tokenization is incorrect.
This explains why it was added in the first place, but I now realize we always have to prefix whatever input we pass to the tokenizer, so there is no point to have this in the user code and risk someone forgetting to do it.

I have verified this using the new Python `test-tokenizer-0.py` by comparing the results with `test-tokenizer-0.cpp`

**Review by SlyEcho — APPROVED** [review]
Seems to be okay with the server, at least.

**Comment by JackJollimore** [comment]
`interactive` in `main` looks good, no issues in my tests.

**Comment by ggerganov** [comment]
@SlyEcho 

I'm looking into more details in this code:

https://github.com/ggerganov/llama.cpp/blob/e4324cbd4d367fb9ddaaef98c0929acebbf22954/examples/server/server.cpp#L298-L307

I think it should be:

```c++
         { 
             if (first) 
             { 
                 prompt_tokens.push_back(bos_token);
                 first = false; 
             } 
             prompt_tokens.push_back(p.template get<llama_token>()); 
         } 
```

Not sure how important is this, but mentioning just in case. You will need to propagate `bos_token` from somewhere though

**Comment by klosax** [comment]
LF token is not correct

```
llm_load_print_meta: general.name   = llama2-7b-hf
llm_load_print_meta: BOS token = 1 '<s>'
llm_load_print_meta: EOS token = 2 '</s>'
llm_load_print_meta: UNK token = 0 '<unk>'
llm_load_print_meta: LF token  = 29871 '▁'
llm_load_tensors: ggml ctx size =    0.09 MB
```

**Comment by JackJollimore** [comment]
> LF token is not correct, should not be tokenized using prepended space.

I missed this, but yeah there's a difference for LF token between PR and master.

**Comment by ggerganov** [comment]
> LF token is not correct, should not be tokenized using prepended space.

Ah good catch. The actual problem is that it should not be tokenized, but we should use `piece_to_id` function

File: examples/perplexity/perplexity.cpp [hunk 2]
```diff
@@ -462,7 +462,13 @@ void hellaswag_score(llama_context * ctx, const gpt_params & params) {
         for (size_t ending_idx = 1; ending_idx < 4; ending_idx++) {
 
             // Tokenize the query
-            query_embd = ::llama_tokenize(ctx, hs_data[task_idx].ending[ending_idx], false);
+            // SPM tokenizer: Do not tokenize the starting space in the ending since it is always added by the tokenizer
+            if (is_spm) {
+                query_embd = ::llama_tokenize(ctx, hs_data[task_idx].ending[ending_idx].substr(1,hs_data[task_idx].ending[ending_idx].size()-1), false);
+            } else {
+                query_embd = ::llama_tokenize(ctx, hs_data[task_idx].ending[ending_idx], false);
+            }
```
Comment (ggerganov) [inline examples/perplexity/perplexity.cpp:470]: I don't think this is correct. We rather have to remove the leading space from here:

https://github.com/ggerganov/llama.cpp/blob/730d9c681e339b76407659344e5a2cd50af7d7d5/examples/perplexity/perplexity.cpp#L395

I've missed that when looking for the places where we manually add leading spaces.

Comment (klosax) [reply examples/perplexity/perplexity.cpp:470]: it is correct and works. the context and endings are concatenated before tokenization further down

Comment (klosax) [reply examples/perplexity/perplexity.cpp:470]: the BPE tokenizer do not add spaces and we need one space

Comment (ggerganov) [reply examples/perplexity/perplexity.cpp:470]: Yes - it gives the correct results, but it is the wrong way to do it. You are special casing the SPM tokenizer this way.
Instead, the general solution is to never prefix the text that you want to tokenize. I don't think any other codebase does something like

```python
text = ' ' + text
tokenizer.encode(text)
```

Regardless of the tokenizer used, the correct code is

```
tokenizer.encode(text)
```

**Comment by klosax** [comment]
Now LF should work in both llama and falcon.

File: examples/perplexity/perplexity.cpp (hunk 2 above)
Comment (klosax) [reply examples/perplexity/perplexity.cpp:470]: When doing hellaswag we need to do this since we need to know exactly which token the endings start. This is where the logprobs we need starts

Comment (klosax) [reply examples/perplexity/perplexity.cpp:470]: We are evaluating the the whole sentences (context+space+ending), but calculate the scores only from the logprobs of the endings.

Comment (klosax) [reply examples/perplexity/perplexity.cpp:470]: A better way would be to have a param to llama_tokenize `prepend_space` like the one to prepend BOS

**Comment by JackJollimore** [comment]
> Now LF should work in both llama and falcon.

llama2 now correctly showing `LF token  = 13 '<0x0A>'` for `main`

File: examples/perplexity/perplexity.cpp (hunk 2 above)
Comment (klosax) [reply examples/perplexity/perplexity.cpp:470]: But we could do it the other way around. Remove the space from the endings and add a space to the BPE instead of subtracting it from the SPM.

**Review by klosax — APPROVED** [review]
Seems to work good.

File: examples/perplexity/perplexity.cpp (hunk 2 above)
Comment (ggerganov) [reply examples/perplexity/perplexity.cpp]: I didn't notice that this branch was before @ikawrakow's fix, so just updated to latest `master`

The #2805 change is the correct way to compute the Hellaswag, because this way we don't implement special logic for different tokenizers

**Review by ikawrakow — APPROVED** [review]

**Comment by ggerganov** [comment]
I have verified that this branch compared to `master` produces:
- the same exact Hellaswag results for F16 Falcon 7B and F16 LLaMA-v2 7B
- identical perplexity for Falcon 7B Q4_0
- nearly identical perplexity for LLaMA-v2 7B Q4_0 (this is expected due to the new correct tokenization of wiki text):
  - `master Q4_0`: ppl 5.9693
  - PR `Q4_0`: ppl 5.9694

**Comment by ikawrakow** [comment]
So, I approved the changes, but the PR basically goes back to auto-adding white space at the beginning strings to be tokenized, which changes HellaSwag scores. I.e., it kind of reverts @klosax's fix in #2806. With PR #2805 merged, this no longer leads to the dramatic difference we saw for HellaSwag after the GGUF merge, but still. I guess, the values in #2321 will need to be updated (for instance, we now have a HellaSwag score of 76.75 after 400 tasks for LLaMA-v2-7B instead of the 77.5 shown in the table in #2321).

**Comment by ggerganov** [comment]
Yes, after digging into this yesterday, I believe that the proposed implementation in this PR is the correct one and the old values have to be updated. The auto-prefix with whitespace inside the `llama_tokenize()` implementation is required in order to match `tokenizer.encode()` behaviour in all cases and strings. So before the PR, we were mostly doing this manually in user-code, but not always. Now we should never need to prefix with whitespace in user-code (unless one wants to concatenate tokenization results, but in this case one must take into account the specific whitespace behaviour of the tokenizer and know how to perform it correctly -- in general, we should avoid concatenation of tokens as we currently do so).

It's still possible that I am misunderstanding some of the tokenizer logic, but atm I'm fairly confident that the proposal is better than what we have on `master` and my unit tests confirm this.

**Comment by ggerganov** [comment]
And one more observation regarding the built-in Falcon BPE tokenization in `llama.cpp`:

We currently know that it does not handle Unicode strings correctly. However, running the unit test, we can see that doing a roundtrip of (tokenize + detokenize) we always get the "source" string:

```bash
$ ./bin/test-tokenizer-0-falcon ../models/falcon-7b/ggml-model-f16.gguf 

src: ' '
res: ' '
tok: 204 

src: '  '
res: '  '
tok: 258 

src: '   '
res: '   '
tok: 466 

src: '    Hello'
res: '    Hello'
tok: 466 23090 

src: '    Hello
    Hello'
res: '    Hello
    Hello'
tok: 466 23090 742 23090 

src: '   Hello'
res: '   Hello'
tok: 258 23090 

src: '  Hello'
res: '  Hello'
tok: 204 23090 

src: ' Hello'
res: ' Hello'
tok: 23090 

src: ' Hello World'
res: ' Hello World'
tok: 23090 2889 

src: ' Hello World!'
res: ' Hello World!'
tok: 23090 2889 12 

src: ' Hello world'
res: ' Hello world'
tok: 23090 1079 

src: ' Hello, world!'
res: ' Hello, world!'
tok: 23090 23 1079 12 

src: ' this is 🦙.cpp'
res: ' this is 🦙.cpp'
tok: 414 304 204 182 237 111 231 25 29247 
main : failed test:    ' this is 🦙.cpp'
main : detokenized to: ' this is 🦙.cpp' instead of ' this is 🦙.cpp'
main : expected tokens:    414,    304,   3346,    111,    231,     25,  29247, 
main : got tokens:         414,    304,    204,    182,    237,    111,    231,     25,  29247, 

src: 'Hello'
res: 'Hello'
tok: 9856 

src: 'Hello World'
res: 'Hello World'
tok: 9856 2889 

src: 'Hello world'
res: 'Hello world'
tok: 9856 1079 

src: 'Hello, world!'
res: 'Hello, world!'
tok: 9856 23 1079 12 

src: 'w048 7tuijk dsdfhu'
res: 'w048 7tuijk dsdfhu'
tok: 98 55866 204 34 16682 7149 36190 6869 11481 

src: 'нещо на Български'
res: 'нещо на Български'
tok: 150 133 6207 151 215 150 134 204 150 133 6279 204 150 223 151 216 15401 150 123 6279 9310 9649 16153 7795 
main : failed test:    'нещо на Български'
main : detokenized to: 'нещо на Български' instead of 'нещо на Български'
main : expected tokens:    150,    133,   6207,    151,    215,    150,    134,   5052,    133,   6279,   5052,    223,    151,    216,  49679,    123,  53110,  47043,   7795, 
main : got tokens:         150,    133,   6207,    151,    215,    150,    134,    204,    150,    133,   6279,    204,    150,    223,    151,    216,  15401,    150,    123,   6279,   9310,   9649,  16153,   7795, 

src: 'កាន់តែពិសេសអាចខលចេញ'
res: 'កាន់តែពិសេសអាចខលចេញ'
tok: 167 236 206 167 236 126 167 236 225 167 237 217 167 236 221 167 237 208 167 236 228 167 236 127 167 236 237 167 237 207 167 236 237 167 236 107 167 236 126 167 236 211 167 236 207 167 236 233 167 236 211 167 237 207 167 236 215 
main : failed test:    'កាន់តែពិសេសអាចខលចេញ'
main : detokenized to: 'កាន់តែពិសេសអាចខលចេញ' instead of 'កាន់តែពិសេសអាចខលចេញ'
main : expected tokens:  38154,    206,  38154,    126,  38154,    225,    167,    237,    217,  38154,    221,    167,    237,    208,  38154,    228,  38154,    127,  38154,    237,    167,    237,    207,  38154,    237,  38154,    107,  38154,    126,  38154,    211,  38154,    207,  38154,    233,  38154,    211,    167,    237,    207,  38154,    215, 
main : got tokens:         167,    236,    206,    167,    236,    126,    167,    236,    225,    167,    237,    217,    167,    236,    221,    167,    237,    208,    167,    236,    228,    167,    236,    127,    167,    236,    237,    167,    237,    207,    167,    236,    237,    167,    236,    107,    167,    236,    126,    167,    236,    211,    167,    236,    207,    167,    236,    233,    167,    236,    211,    167,    237,    207,    167,    236,    215, 

src: '🚀 (normal) 😶‍🌫️ (multiple emojis concatenated) ✅ (only emoji that has its own token)'
res: '🚀 (normal) 😶‍🌫️ (multiple emojis concatenated) ✅ (only emoji that has its own token)'
tok: 182 237 232 206 204 19 11003 20 204 182 237 230 126 168 206 219 182 237 218 116 13392 204 19 51831 732 63209 1741 7955 522 20 204 57196 204 19 7927 53360 325 504 701 946 10930 20 
main : failed test:    '🚀 (normal) 😶‍🌫️ (multiple emojis concatenated) ✅ (only emoji that has its own token)'
main : detokenized to: '🚀 (normal) 😶‍🌫️ (multiple emojis concatenated) ✅ (only emoji that has its own token)' instead of '🚀 (normal) 😶‍🌫️ (multiple emojis concatenated) ✅ (only emoji that has its own token)'
main : expected tokens:   2571,    232,    206,    204,     19,  11003,     20,   8196,    126,    283,    219,  48778,    116,  13392,    204,     19,  51831,    732,  63209,   1741,   7955,    522,     20,  22438,    211,    204,     19,   7927,  53360,    325,    504,    701,    946,  10930,     20, 
main : got tokens:         182,    237,    232,    206,    204,     19,  11003,     20,    204,    182,    237,    230,    126,    168,    206,    219,    182,    237,    218,    116,  13392,    204,     19,  51831,    732,  63209,   1741,   7955,    522,     20,    204,  57196,    204,     19,   7927,  53360,    325,    504,    701,    946,  10930,     20, 
```

However, the obtained string is sometimes constructed from a different set of tokens that does not match the correct set that we obtain with the OG tokenizer. This happens when unicode characters are involved.

All of this is expected and indicates that for now, projects should avoid `llama.cpp`'s built-in BPE tokenizer if tokenization accuracy is important. But for quick tests, the generated texts should look OK

**Comment by ikawrakow** [comment]
Btw., did anyone except me notice that tokenization is now almost 6 times slower than it used to be?

After merging this PR, it now takes 3.2 seconds to tokenize Wikitext. It used to be ~540 ms up to bae5c5f679e043371bc2b4dffff8d4964d6cb953, became 3.1 seconds after merging #2806, and now 3.2 seconds after this PR.

**Comment by JackJollimore** [comment]
I don't have an accurate measure for timing for tokenizing input in `perplexity`, but counting in my head took 23 seconds for both master & https://github.com/ggerganov/llama.cpp/commit/bae5c5f679e043371bc2b4dffff8d4964d6cb953

I'd notice a slowdown **a lot** on Android unless my device is unaffected.

**Comment by OthmanProgramming** [comment]
I'm need solve this model please
Min F(S_1,S_2,S_3 ):850 S_1^2+825S_2^2+660S_3^2                 
s.to
1650S_1-μ_1 (1000000S_1 )-μ_2 (30.464S_1 )-μ_3 (24S_1 )-μ_4 (200000S_1 )-μ_5 (1800S_1 )=0                                                                                        (1)   
 1600S_2-μ_1 (87300S_2 )-μ_2 (24.6421S_2 )-μ_3 (16S_2 )-μ_4 (200000S_2 )
                  -μ_5 (1700S_2)=0                                                                                         (2)   
1200S_3-μ_1 (772800S_3 )-μ_2 (19.944S_3 )-μ_3 (10S_3 )-μ_4 (16000S_3 )
                              -μ_5 (1320 S_3 )=0                                                                                (3)   
500000S_1^2+436500S_2^2+386400S_3^2     ≤2322900                                           (4)  
15.232S_1^2+12.321S_2^2+9.972S_3^2                 ≤400000                                             (5)  
12S_1^2+8S_2^2+5 S_3^2                                           ≤2000                                                   (6)  
100000S_1^2+100000S_2^2+800000S_3^2       ≤500000                                             (7)  
900S_1^2+850S_2^2+660S_3^2                                 ≤7650000                                              (8)
 μ_1 (500000S_1^2+436500S_2^2+386400S_1^2-2322900)=0                             (9)
 μ_2 (15.232S_1^2+12.321S_2^2+9.972S_1^2  -400000)            =0                            (10)
μ_3 (12S_1^2+8S_2^2+5S_3^2-16000000)                                   =0                               (11)
μ_4 (100000S_1^2+100000S_2^2+800000S_3^2-500000)   =0                               (12)
 μ_5 (  900S_1^2+850S_2^2+660 S_3^2  -7650000)                       =0                                                (13)
 μ_5,μ_2,μ_3,μ_4,μ_5                                                                        ≤0                                                                  (14)  
     S_1,S_2,S_3                                                                                        ≥0                                                                  (15)
                                                  S1, S2, S3≥0


solve in python

**Comment by OthmanProgramming** [comment]
I'm need solve this model
Min F(S_1,S_2,S_3 ):850 S_1^2+825S_2^2+660S_3^2                 
s.to
1650S_1-μ_1 (1000000S_1 )-μ_2 (30.464S_1 )-μ_3 (24S_1 )-μ_4 (200000S_1 )-μ_5 (1800S_1 )=0                                                                                        (1)   
 1600S_2-μ_1 (87300S_2 )-μ_2 (24.6421S_2 )-μ_3 (16S_2 )-μ_4 (200000S_2 )
                  -μ_5 (1700S_2)=0                                                                                         (2)   
1200S_3-μ_1 (772800S_3 )-μ_2 (19.944S_3 )-μ_3 (10S_3 )-μ_4 (16000S_3 )
                              -μ_5 (1320 S_3 )=0                                                                                (3)   
500000S_1^2+436500S_2^2+386400S_3^2     ≤2322900                                           (4)  
15.232S_1^2+12.321S_2^2+9.972S_3^2                 ≤400000                                             (5)  
12S_1^2+8S_2^2+5 S_3^2                                           ≤2000                                                   (6)  
100000S_1^2+100000S_2^2+800000S_3^2       ≤500000                                             (7)  
900S_1^2+850S_2^2+660S_3^2                                 ≤7650000                                              (8)
 μ_1 (500000S_1^2+436500S_2^2+386400S_1^2-2322900)=0                             (9)
 μ_2 (15.232S_1^2+12.321S_2^2+9.972S_1^2  -400000)            =0                            (10)
μ_3 (12S_1^2+8S_2^2+5S_3^2-16000000)                                   =0                               (11)
μ_4 (100000S_1^2+100000S_2^2+800000S_3^2-500000)   =0                               (12)
 μ_5 (  900S_1^2+850S_2^2+660 S_3^2  -7650000)                       =0                               (13)
      μ_5,μ_2,μ_3,μ_4,μ_5                                                                           ≤0                             (14)  
     S_1,S_2,S_3                                                                                        ≥0                              (15)
                                                  S1, S2, S3≥0


solve in python

**Comment by shibe2** [comment]
So how do I tokenize partial text now? I need `tokenize(x+y)==tokenize(x)+tokenize(y)`.

**Comment by goerch** [comment]
> So how do I tokenize partial text for Lllama 2 now? I need `tokenize(x+y)==tokenize(x)+tokenize(y)` (given that ending of x and beginning of y don't form a token).

Interesting. This is an invariant I haven't looked into yet. For [now](https://github.com/ggerganov/llama.cpp/issues/2310#issuecomment-1705509602):

```
        for(auto i = 0; i < 1024; ++ i) {
            auto str = random_string(64);
            for(auto j = 0; j < str.length(); ++ j) {
                auto ids1 = spp.EncodeAsIds(str.substr(0, j));
                auto ids2 = spp.EncodeAsIds(str.substr(j));
                auto str1 = spp.DecodeIds(ids1);
                auto str2 = spp.DecodeIds(ids2);
                if (str1 + str2 != str)
                    std::cout << "1: >" << str1 << "< >" << str2 << "< >" << str << "<" << std::endl;
            }
        }
        for(auto i = 0; i < 1024; ++ i) {
            auto str = random_string(64);
            auto ids = spp.EncodeAsIds(str);
            for(auto j = 0; j < ids.size(); ++j) {
                auto str1 = spp.DecodeIds(std::vector(ids.begin(), ids.begin() + j));
                auto str2 = spp.DecodeIds(std::vector(ids.begin() + j, ids.end()));
                if (str1 + str2 != str)
                    std::cout << "2: >" << str1 << "< >" << str2 << "< >" << str << "<" << std::endl;
            }
        }
```

Invariant 1 seems to hold for Meta's tokenizer, invariant 2 breaks as soon as spaces are involved.

**Comment by goerch** [comment]
If my quick check 

```
        for(auto i = 0; i < 1024; ++ i) {
            auto str = random_string(64);
            auto ids = spp.EncodeAsIds(str);
            for(auto j = 0; j < str.length(); ++ j) {
                auto ids1 = spp.EncodeAsIds(str.substr(0, j));
                auto ids2 = spp.EncodeAsIds(str.substr(j));
                auto check = ids1;
                check.insert(check.end(), ids2.begin(), ids2.end());
                if (check != ids) {
                    auto str1 = spp.DecodeIds(ids1);
                    auto str2 = spp.DecodeIds(ids2);
                    std::cout << "3: >" << str1 << "< >" << str2 << "< >" << str << "<" << std::endl;
                }
            }
        }
```

is correct, this invariant also doesn't hold for Meta's tokenizer in the presence of spaces.

**Comment by shibe2** [comment]
> So how do I tokenize partial text for Lllama 2 now?

I have found a way. I prefix the string with LF, then remove tokens up to and including the first `llama_token_nl`. I added a parameter to server /tokenize, which triggers this behavior (only effective for `LLAMA_VOCAB_TYPE_SPM`).

By the way, server documentation as of now says that this should be the default behavior:

> Note that the special BOS token is not added in front of the text and also a space character is not inserted automatically as it is for /completion.
